src/libcore/array.rs

```diff
@@ -92,6 +96,275 @@ where
+        // `data[alive.start..alive.end + 1]`. Reading this element means that
+        // `data[alive.end]` is regarded as dead now (i.e. do not touch). As
+        // `alive.end` was the end of the alive-zone, the alive zone is now
+        // `data[alive]` again, restoring all invariants.
+        let out = unsafe { self.data.get_unchecked(self.alive.end).read() };
+
+        Some(out)
+    }
+}
+
+#[unstable(feature = "array_value_iter", issue = "0")]
+impl<T, const N: usize> Drop for IntoIter<T, {N}>
+where
+    [T; N]: LengthAtMost32,
+{
+    fn drop(&mut self) {
+        // Here we drop all alive elements (but nothing else). It is safe to
+        // call `as_mut_ptr` as all elements within the alive range are
+        // correctly initialized.
+        //
+        // We increment `alive.start` before dropping for exception safety. If
+        // the `drop` panics, we rather want to treat that element as already
+        // droped (with the risk of memory leakage) instead of treating it as
+        // undropped (with the risk of memory unsafety, which is far worse).
+        unsafe {
+            while self.alive.start != self.alive.end {
+                let idx = self.alive.start;
+                self.alive.start += 1;
+                ptr::drop_in_place(self.data.get_unchecked_mut(idx).as_mut_ptr());
+            }
+        }
+    }
+}
+
+#[unstable(feature = "array_value_iter", issue = "0")]
+impl<T, const N: usize> ExactSizeIterator for IntoIter<T, {N}>
+where
+    [T; N]: LengthAtMost32,
+{
+    fn len(&self) -> usize {
+        self.len_remaining()
+    }
+    fn is_empty(&self) -> bool {
+        self.alive.is_empty()
+    }
+}
+
+#[unstable(feature = "array_value_iter", issue = "0")]
+impl<T, const N: usize> FusedIterator for IntoIter<T, {N}>
+where
+    [T; N]: LengthAtMost32,
+{}
+
+// The iterator indeed reports the correct length. The number of "alive"
+// elements (that will still be yielded) is the length of the range `alive`.
+// This range is decremented in length in either `next` or `next_back`. It is
+// always decremented by 1 in those methods, but only if `Some(_)` is returned.
+#[unstable(feature = "array_value_iter", issue = "0")]
+unsafe impl<T, const N: usize> TrustedLen for IntoIter<T, {N}>
+where
+    [T; N]: LengthAtMost32,
+{}
+
+#[unstable(feature = "array_value_iter", issue = "0")]
+impl<T: Clone, const N: usize> Clone for IntoIter<T, {N}>
+where
+    [T; N]: LengthAtMost32,
+{
+    fn clone(&self) -> Self {
+        unsafe {
+            // This creates a new uninitialized array. Note that the `assume_init`
+            // refers to the array, not the individual elements. And it is Ok if
+            // the array is in an uninitialized state as all elements may be
+            // uninitialized (all bit patterns are valid). Compare the
+            // `MaybeUninit` docs for more information.
+            let mut new_data: [MaybeUninit<T>; N] = MaybeUninit::uninit().assume_init();
```

Comment:
Use `[MaybeUninit::UNINIT; N]` and hoist `let mut data` out of the `unsafe {` block?

Reply:
```
error[E0277]: the trait bound `T: marker::Copy` is not satisfied
   --> src/libcore/array.rs:330:49
    |
330 |         let mut new_data: [MaybeUninit<T>; N] = [MaybeUninit::UNINIT; N];
    |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `marker::Copy` is not implemented for `T`
```

But apart from that, I put the whole body in the unsafe block because all operations are closely linked basically. I'm a fan of sometimes using larger `unsafe` scopes to include the technically safe stuff which still has an effect on the unsafe stuff.

Reply:
@LukasKalbertodt hmm; you probably need the feature gate in `libcore/lib.rs` then. (I see no test in the impl PR that actually mentions needing the gate... cc @davidtwco)

Reply:
I'm afraid I don't understand. Are you talking about [this feature](https://github.com/rust-lang/rust/issues/49147)? So you absolutely want to have it changed? If you don't have a strong opinion about this, I'd keep it as is.

Reply:
> Are you talking about [this feature](https://github.com/rust-lang/rust/issues/49147)?

I am; See https://github.com/rust-lang/rust/pull/62799 for an example of where it is used.

> So you absolutely want to have it changed? If you don't have a strong opinion about this, I'd keep it as is.

I mean I don't insist on that it must happen now but the feature gate is intended to make this nicer among others and it avoids one `unsafe { ... }` entirely so we should use it if we can. This seems like as good a time as any.

Reply:
> `#![cfg_attr(not(bootstrap), feature(const_in_array_repeat_expressions))]`

The iterator (my changes) is currently also compiled with the bootstrap compiler, so we:
- could wait for the bootstrap compiler to be updated, or
- add both versions (current and your suggestion) and select them via `cfg`, or
- keep it as is but add a FIXME comment to change it later once the bootstrap compiler is updated.

What do you think?

Reply:
> add both versions (current and your suggestion) and select them via `cfg`, or

This option sounds best since it will guarantee that Mark will fix it once we cut a new release & update bootstrap. :)

Reply:
Does this new iterator _need_ to be available in bootstrap?  Could the whole type just be cfg'd-out in bootstrap?  (That also isolates it from any problems with const generics in the bootstrap compiler...)

Reply:
If so I'd prefer putting the new stuff in its own module which is probably for the best anyways in terms of isolating the unsafe code and also for having better separation of concerns.

Reply:
Locally (not pushed yet) I created a new module for the iterator stuff and guarded it with `#[cfg(not(bootstrap))]`. As this PR is still blocked by const generic bugs, it makes sense avoid waiting for the bootstrap compiler to update.

But regarding the original remark: `[MaybeUninit::UNINIT; N]` now leads to:

```
error: array lengths can't depend on generic parameters
   --> src/libcore/array/iter.rs:233:50
    |
233 |         let mut new_data = [MaybeUninit::UNINIT; {N}];
    |                                                  ^^^
```

Apparently this is a known limitation?

Reply:
@LukasKalbertodt Interesting; This problem seems to arise due to the anonymous constant not having `N` in scope as the usual limitation wrt. not having [scoped type variables](https://www.microsoft.com/en-us/research/publication/lexically-scoped-type-variables/).

I tried a work-around:

```rust
#![feature(const_generics)]

struct A<const N: usize>;

use core::mem::MaybeUninit;

impl<const N: usize> A<{N}> {
    fn foo() {
        let x: [MaybeUninit<u8>; N] = [MaybeUninit::UNINIT; _];
    }
}
```
but I forgot that `_` is not a valid expression form (it should be!).

cc @varkor -- any ideas? if not we'll have to use the `unsafe { ... }` for now.